Atru/resource usage pages (#719)
hashicorp/web-unified-docs: content/terraform/v1.12.x/docs/language/resources/configure.mdx

---

# Configure a resource

A **resource** is any infrastructure object you want to create and manage with Terraform, including virtual networks, compute instances, or higher-level components such as DNS records. To create resources in Terraform, add `resource` blocks to your Terraform configuration and specify the required arguments. 

This topic provides information about creating and managing resources using `resource` blocks. For an overview of how to provision infrastructure, refer to [Core Terraform workflow](/terraform/intro/core-workflow). For details about configuring a `resource` block, refer to your provider documentation and the [`resource` block reference](/terraform/language/block/resource). The `resource` block reference describes the native Terraform [meta-arguments](/terraform/language/meta-arguments) you can use to configure how Terraform creates a resource and manages its lifecycle.

> **Hands-on:** Try the [Terraform: Get Started](/terraform/tutorials/aws-get-started?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorials.

## Overview

Complete the following steps to configure a resource:

> - Declare a resource and specify its type and a name.

Reviewer: ```suggestion
- Declare a resource block and specify its type and a name.
```

Maybe?

## Prerequisites

Choose and configure your provider. Providers are plugins for Terraform that offer a collection of resource types that you can provision. Refer to [Providers](/terraform/language/providers) for more information. 

Most providers require authentication settings, version requirements, or other arguments to access their remote APIs. Terraform expects provider configuration in the root module. Refer to the [`provider` block reference](/terraform/language/block/provider) for details.

To manage resources, a Terraform module must specify the required providers. Refer to [Provider requirements](/terraform/language/providers/requirements) for more information.

## Declare a resource

Add a `resource` block to your configuration, declare a type, and specify a name for it. The type determines the kind of infrastructure object the `resource` block manages and what arguments the resource supports. Most resource types are implemented by your provider, but Terraform includes a native resource type, [`terraform_data`](/terraform/language/resources/terraform-data), that you can use to create resources for specialized use cases. 

> You must also specify a local name for the resource. Terraform uses the resource type and name as the resource's address in the [state file](/terraform/language/state). 

Reviewer: ```suggestion
You must also specify a local name for the resource. Terraform uses the resource type and name as the resource's address in your workspace's [State](/terraform/language/state). 
```

Matching the page title.

## Configure provider-specific arguments

Most of the arguments within the body of a `resource` block are specific to the selected resource type as defined by the provider developer. Configure your resource by assigning values to the `resource` block’s arguments in your configuration arguments. The resource type's documentation lists which arguments are required and how to format their values. Refer to your provider documentation for details. 

> You can use Terraform [expressions](/terraform/language/expressions) and other dynamic Terraform language features to configure resource argument values in addition to hard-coding values.   

Reviewer: ```suggestion
You can use Terraform [Expressions](/terraform/language/expressions) and other dynamic Terraform language features to configure resource argument values in addition to hard-coding values.   
```

I *think* we match the page title, including case (other examples in this PR do it that way), but I can't find specific guidance on this in the style guide.

Terraform providers write and maintain their own documentation. For all publicly-available provider documentation, refer to the public [Terraform registry](https://registry.terraform.io/browse/providers).

> ### Define operation timeouts

Author: Noted, but per our conversation, provider docs in the registry link to it, so we need to keep it here for now.

> ### Define operation timeouts

Reviewer: Could/should we move it further down the page? It's probably the least-used "advanced" concept on this page.

> ### Define operation timeouts

Author: This describes a provider-specific argument, so I don't think it would make sense (or would make less sense) to move it elsewhere.

You can use the following methods to control the order that Terraform follows to create resources:

- Add the [`depends_on` meta-argument](/terraform/language/block/resource#depends_on) to the `resource` block and reference the upstream resource that Terraform must create first. For an example configuration, refer to [Specify a dependency](/terraform/language/block/resource#specify-a-dependency).

- Add [`replace_triggered_by` lifecycle rule](/terraform/language/block/resource#replace_triggered_by) to the `resource` block.  This argument adds dependencies between independent resources by forcing Terraform to replace the parent resource when there is a change to a referenced resource or resource attribute. For an example configuration, refer to [Specify triggers that replace resources](/terraform/language/block/resource#specify-triggers-that-replace-resources).

## Configure built-in resources

You can declare special resource types that help you program actions without affecting actual infrastructure.

### `terraform_data` resources

> The `terraform_data` resource type implements the standard resource lifecycle, but does not directly take any other actions. Refer to [`terraform_data` resource reference](/terraform/language/resources/terraform_data) for more information.

Author: ```suggestion
The `terraform_data` resource type implements the standard resource lifecycle, but does not directly take any other actions. Refer to [`terraform_data` resource reference](/terraform/language/resources/terraform-data) for more information.
```
